src/flag_gems/ops/isin.py

```diff
@@ -0,0 +1,259 @@
+        block = tl.where(
+            mask,
+            tl.where(invert, block and (in0 != in1), block or (in0 == in1)),
+            invert,
+        )
+    out = tl.reduce(block, axis=1, combine_fn=(reduce_all if invert else reduce_any))
+    tl.store(out_ptr, out[:, None], row_mask)
+
+
+@libentry()
+@triton.jit
+def isin_by_comparation_kernel(
+    in0_ravel_ptr: tl.tensor,
+    in1_ravel_ptr: tl.tensor,  # in
+    out_ptr: tl.tensor,  # out
+    M: int,  # num_tasks
+    N: int,  # num_tasks_1
+    BLOCK_M: tl.constexpr,  # tile_size
+    BLOCK_N: tl.constexpr,  # tile_size_1
+    tiles_per_cta: int,
+    invert: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    num_ctas = tl.num_programs(0)
+    # grid-stride-loop style kernel
+    for j in range(0, tiles_per_cta):
+        global_pid = pid + j * num_ctas if j > 0 else pid
+        isin_by_comparation_impl(
+            global_pid,
+            in0_ravel_ptr,
+            in1_ravel_ptr,  # in
+            out_ptr,  # out
+            M,
+            N,
+            BLOCK_M,
+            BLOCK_N,
+            invert,
+        )
+
+
+def isin_by_comparation(
+    in0: torch.tensor,
+    in1: torch.tensor,
+    invert: bool,
+):
+    in0_ravel = in0.contiguous().ravel()
+    in1_ravel = in1.contiguous().ravel()
+    M = in0.numel()
+    N = in1.numel()
+    if M <= 1024:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(1, 256, N, 4)
+    elif M <= 3072:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(2, 256, N, 4)
+    elif M <= 6144:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(4, 128, N, 4)
+    elif M <= 9216:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(4, 256, N, 8)
+    else:
+        BLOCK_M, BLOCK_N, num_warps = launch_arg(4, 128, N, 4)
+    num_ctas = min(65536, triton.cdiv(M, BLOCK_M))
+    tiles_per_cta = triton.cdiv(M, BLOCK_M * num_ctas)
+    grid = (num_ctas,)
+    out = torch.empty_like(in0_ravel, dtype=torch.bool)
+    with torch.cuda.device(in0_ravel.device.index):
+        isin_by_comparation_kernel[grid](
+            in0_ravel,
+            in1_ravel,  # in
+            out,  # out
+            M,
+            N,
+            BLOCK_M,
+            BLOCK_N,
+            tiles_per_cta=tiles_per_cta,
+            invert=invert,
+            num_warps=num_warps,
+        )
+    return out.view_as(in0)
+
+
+@triton.jit
+def isin_by_search_impl(
+    global_pid,
+    in0_ravel_ptr: tl.tensor,
+    in1_sorted_ptr: tl.tensor,  # in
+    out_ptr: tl.tensor,  # out
+    M: int,  # num_tasks
+    N: int,  # num_tasks_1
+    BLOCK_M: tl.constexpr,  # tile_size
+    invert: tl.constexpr,
+):
+    r = tl.arange(0, BLOCK_M)
+    i0 = global_pid * BLOCK_M + r
+    mask = i0 < M
+
+    # load in0_ravel
+    in0_ravel = tl.load(in0_ravel_ptr + i0, mask=mask)
+
+    # binary search: lower_bound
+    out = tl.zeros_like(r).to(tl.int1)
+    start = tl.zeros_like(r)
+    end = start + N
+    while_mask = start < end
+    while tl.sum(tl.where(while_mask, out, 1)) != BLOCK_M:
```

Comment:
To save an extra reduction at the end of each iteration, I would rather use a fixed loop bound log2(N) + 1 and make sure that start and end meet in the end.

Reply:
> To save an extra reduction at the end of each iteration, I would rather use a fixed loop bound log2(N) + 1 and make sure that start and end meet in the end.

done